### PR TITLE
GaroRobe/issue1150

### PR DESCRIPTION
**Pull request by GaroRobe**
### Description of the Change

Introduced KURA_BLOCKS_PER_STORAGE_FILE setting which defaults to 1000 and defines number of blocks per each new created datafile. 

### Benefits

1. Persistent storage logic was reduced from "random block read/write" to "bulk read+decoding/encoding+append". It should significantly speed up read times (which is to be measured!) without any risks on write (append write of single 4Kb IO Block is usually atomic).
2. Within data file we store blocks as rCOBS-encoded 0u8-terminated bitspans. As a result, we don't rely on number of blocks in file for deserialization and can reconfigure this setting for node at any moment without consequences.

### Possible Drawbacks 

Since file format was modified, this change is breaking backwards compatibility

### Usage Examples or Tests *[optional]*

I have changed test read_all_blocks_from_block_store accordingly.

### Alternate Designs *[optional]*

Embedded DB is a good alternative, and definitely IS a better alternative if we'll eventually need random access to serialized blocks.

### CODE REVIEW

File: iroha/Cargo.toml [hunk 1]
```diff
@@ -41,6 +41,7 @@ iroha_schema = { path = "../iroha_schema"}
 
 async-trait = "0.1"
 chrono = "0.4"
+rcobs = "0.1"
```
Comment (i1i1) [inline iroha/Cargo.toml:44]: Do we need `rcobs` here? Can we do stream writing manually?

File: iroha/src/kura.rs [hunk 2]
```diff
@@ -180,77 +180,129 @@ impl Default for Mode {
...
+    blocks_per_file: u64,
 }
 
 impl BlockStore {
     /// Initialize block storage at `path`.
     ///
     /// # Errors
     /// - Failed to create directory.
-    pub fn new(path: &Path) -> Result<BlockStore> {
+    pub fn new(path: &Path, blocks_per_file: u64) -> Result<BlockStore> {
```
Comment (i1i1) [inline iroha/src/kura.rs:191]: `NonZero<u64>` would suit better i think. Also you can add second constructor for defaulting `blocks_per_file`

File: iroha/src/kura.rs [hunk 3]
```diff
@@ -273,13 +326,16 @@ pub mod config {
         /// Path to the existing block store folder or path to create new folder.
         #[serde(default = "default_kura_block_store_path")]
         pub kura_block_store_path: String,
+        #[serde(default = "default_kura_blocks_per_storage_file")]
+        pub kura_blocks_per_storage_file: u64,
```
Comment (i1i1) [inline iroha/src/kura.rs:330]: Could you remove prefix kura from fields? Also add this after `serde` rename_all before struct:
```rust
#[config(env_prefix = "KURA_")]
```

File: iroha/src/kura.rs [hunk 4]
```diff
@@ -1,18 +1,15 @@
 //! This module contains persistence related Iroha logic.
 //! [`Kura`] is the main entity which should be used to store new [`Block`](`crate::block::VersionedCommittedBlock`)s on the blockchain.
 
-use std::{
-    fs,
-    path::{Path, PathBuf},
-    sync::Arc,
-};
+use std::{fs::{self}, path::{Path, PathBuf}, sync::Arc};
```
Comment (i1i1) [inline iroha/src/kura.rs:4]: Could you try not to mix async and sync api. It is bad generally

File: iroha/src/kura.rs [hunk 5]
```diff
@@ -180,77 +180,129 @@ impl Default for Mode {
...
 
-    fn get_block_filename(block_height: u64) -> String {
-        format!("{}", block_height)
+    /// Returns expected filename of datafile containing block, provided its height
+    async fn get_block_filename(dir: &PathBuf, block_height: u64, blocks_per_file: u64) -> Result<String> {
+        let storage_indices = sorted_storage_files_base_indices(dir).await
+            .wrap_err("Failed to list storage files")?;
+        match storage_indices.last() {
+            Some(max_base) => {
+                assert!(block_height > *max_base);
```
Comment (i1i1) [inline iroha/src/kura.rs:207]:
```suggestion
                debug_assert!(block_height > *max_base);
```

Comment (i1i1) [reply iroha/src/kura.rs:207]: As `assert!` will be in release build also, while debug will be compiled out

File: iroha/src/kura.rs [hunk 6]
```diff
@@ -180,77 +180,129 @@ impl Default for Mode {
...
-            height += 1;
+        let mut blocks: Vec<VersionedCommittedBlock> = Vec::new();
+        let base_indices = sorted_storage_files_base_indices(&self.path).await.unwrap();
+        for base_index in base_indices {
+            let path = self.path.join(base_index.to_string());
+            let mut read_blocks = read_all_from_storage_file(&path).await.expect("Failed to deserialize blocks.");
+            if !blocks.is_empty() {
+                assert_eq!(read_blocks.first().unwrap().header().height, blocks.last().unwrap().header().height + 1);
+            } else {
+                assert_eq!(read_blocks.first().unwrap().header().height, 1);
```
Comment (i1i1) [inline iroha/src/kura.rs:255]: Same for these asserts.

File: iroha/src/kura.rs [hunk 7]
```diff
@@ -180,77 +180,129 @@ impl Default for Mode {
...
     /// # Errors
     /// - Failed to create directory.
-    pub fn new(path: &Path) -> Result<BlockStore> {
+    pub fn new(path: &Path, blocks_per_file: u64) -> Result<BlockStore> {
         if fs::read_dir(path).is_err() {
             fs::create_dir_all(path).wrap_err("Failed to create Block Store directory.")?;
         }
         Ok(BlockStore {
             path: path.to_path_buf(),
+            blocks_per_file: if blocks_per_file == 0 { config::DEFAULT_BLOCKS_PER_STORAGE_FILE } else { blocks_per_file },
```
Comment (i1i1) [inline iroha/src/kura.rs:197]: It is bad api in rust to encode one of integer variants as special thing. Enums suit better for that.

**Review by i1i1 — CHANGES_REQUESTED** [review]

**Comment by GaroRobe** [comment]
> Do we need `rcobs` here? Can we do stream writing manually? 

We could. But I had 2 things in mind:
1. It should be easy to parse
2. It should be possible to add new blocks by appending them at the end of file

(r)COBS is something I used recently for exact this purpose. How will it be done using stream write?

**Comment by GaroRobe** [comment]
> `NonZero<u64>` would suit better i think. Also you can add second constructor for defaulting `blocks_per_file` 

Good points, I somehow didn't think of NonZero. Thanks

**Comment by GaroRobe** [comment]
> Could you remove prefix kura from fields? Also add this after `serde` rename_all before struct:
```rust
#[config(env_prefix = "KURA_")]
``` 

Is it related to kura_block_store_path as well?

**Comment by GaroRobe** [comment]
> As `assert!` will be in release build also, while debug will be compiled out 

Is it not something critical in release? How should I rather report an error here?

File: iroha/Cargo.toml (hunk 1 above)
Comment (GaroRobe) [reply iroha/Cargo.toml:44]: We could. I will do that.

File: iroha/src/kura.rs (hunk 2 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:191]: Good points, I somehow didn't think of NonZero. Thanks

File: iroha/src/kura.rs (hunk 3 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:330]: Is it related to kura_block_store_path as well?

File: iroha/src/kura.rs (hunk 2 above)
Comment (i1i1) [reply iroha/src/kura.rs:191]: General intention is to use Correct-by-construction pattern. Do not validate input, use types for type validation (types which are have some properties (like `NonZero`))

File: iroha/src/kura.rs (hunk 7 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:197]: NonZero as well then?

File: iroha/src/kura.rs (hunk 3 above)
Comment (i1i1) [reply iroha/src/kura.rs:330]: Yep, to all fields. It used for setting fields in configuration from shell environment variables.

File: iroha/src/kura.rs (hunk 5 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:207]: So, should I use Result as a return type? Or just truncate successfull deseralization streak and log error? Which is more appropriate?

Comment (eadventurous) [reply iroha/src/kura.rs:207]: Actually yes better to use Result than asserts.
Create `BlockStoreError` enum with variants like `InconsistentHeight(...)`, `FailedToRead` ...
And then return Result<_, BlockStoreError>

Comment (eadventurous) [reply iroha/src/kura.rs:207]: I think your point about not truncating but failing for ease of administration - is valid

File: iroha/src/kura.rs (hunk 4 above)
Comment (eadventurous) [reply iroha/src/kura.rs:4]: Use tokio::fs and etc.

File: iroha/src/kura.rs (hunk 5 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:207]: Then I will adhere to this logic:
1. Managed to deserialized - success
2. Anything went wrong - report an error

File: iroha/src/kura.rs [hunk 8]
```diff
@@ -180,126 +189,252 @@ impl Default for Mode {
...
     }
 
-    fn default_kura_block_store_path() -> String {
-        DEFAULT_KURA_BLOCK_STORE_PATH.to_owned()
+    fn default_block_store_path() -> String {
+        DEFAULT_BLOCK_STORE_PATH.to_owned()
+    }
+
+    fn default_blocks_per_storage_file() -> NonZeroU64 {
+        #![allow(clippy::expect_used, clippy::restriction)]
```
Comment (GaroRobe) [inline iroha/src/kura.rs:434]: Ok, here's something I really want to consult about:
I think forcing clippy here is ugly, but simply can't find a better way. This error is not possible by design, the only way we could get it is by having typo-change in constant. So, is it ok to mute clippy for such a case?

File: iroha/src/kura.rs [hunk 9]
```diff
@@ -180,126 +189,252 @@ impl Default for Mode {
...
+            None => Ok(String::from(INITIAL_DATA_FILE)),
+        }
     }
 
     /// Get filysystem path for the block at height.
-    pub fn get_block_path(&self, block_height: u64) -> PathBuf {
-        self.path.join(BlockStore::get_block_filename(block_height))
+    ///
+    /// # Errors
+    ///
```
Comment (i1i1) [inline iroha/src/kura.rs:237]: Could you add comments here?

File: iroha/src/kura.rs [hunk 10]
```diff
@@ -180,126 +189,252 @@ impl Default for Mode {
...
     /// Write block as file onto the disk.
     #[iroha_futures::telemetry_future]
     pub async fn write(&self, block: &VersionedCommittedBlock) -> Result<Hash> {
-        //filename is its height
-        let path = self.get_block_path(block.header().height);
-        let mut file = File::create(path)
+        let path = self
+            .get_block_path(
+                NonZeroU64::new(block.header().height)
+                    .ok_or_else(|| eyre::eyre!("Block height is non-positive."))?,
```
Comment (i1i1) [inline iroha/src/kura.rs:252]:
```suggestion
                    .ok_or_else(|| eyre::eyre!("Block height is zero"))?,
```

File: iroha/src/kura.rs [hunk 11]
```diff
@@ -180,126 +189,252 @@ impl Default for Mode {
...
+            .create(true)
+            .append(true)
+            .open(path)
             .await
             .wrap_err("Failed to open storage file.")?;
         let hash = block.hash();
-        let serialized_block: Vec<u8> = block.encode_versioned()?;
+        let mut encoded: Vec<u8> = block.encode_versioned()?;
+        let serialized_block = prepend_size(&mut encoded);
         file.write_all(&serialized_block)
```
Comment (i1i1) [inline iroha/src/kura.rs:264]: You can write size directly to file and then write encoded block. That should do the job.

File: iroha/src/kura.rs [hunk 12]
```diff
@@ -180,126 +189,252 @@ impl Default for Mode {
...
-        while let Ok(block) = self.read(height).await {
-            blocks.push(block);
-            height += 1;
+fn prepend_size(data: &mut [u8]) -> Vec<u8> {
+    let mut result = Vec::new();
+    result.reserve(data.len() + std::mem::size_of::<usize>());
+    result.extend_from_slice(&data.len().to_ne_bytes());
+    result.extend_from_slice(data);
+    result
+}
```
Comment (i1i1) [inline iroha/src/kura.rs:311]: `std::mem::size_of::<usize>()` is platform dependent. It is either 4 or 8. Change it to `u32` or `u64`

File: iroha/src/kura.rs (hunk 8 above)
Comment (i1i1) [reply iroha/src/kura.rs:434]: Thats ok here. You shouldn't allow lint groups (for example `clippy::restriction`) though

Generally using unwraps or expects is fine if you justify that they are unreachable.

File: iroha/src/kura.rs (hunk 9 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:237]: Done

File: iroha/src/kura.rs (hunk 10 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:252]: Done

File: iroha/src/kura.rs (hunk 11 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:264]: You dislike another function here?

File: iroha/src/kura.rs (hunk 12 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:311]: But usize is system-dependent - 2,4 or 8 bytes (so far)
I did it this way specifically because I expect that I don't know target system, but expect that datafiles never migrate between platforms.

File: iroha/src/kura.rs (hunk 11 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:264]: You dislike a new function here?

File: iroha/src/kura.rs (hunk 12 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:311]: But usize is system-dependent - 2,4 or 8 bytes (so far)
I did it this way specifically because I expect that I don't know target system, but expect that datafiles never migrate between platforms.

Comment (GaroRobe) [reply iroha/src/kura.rs]: Resolved by setting size to u32 explicitly.

File: iroha/src/kura.rs (hunk 11 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:269]: Removed it.

File: iroha/src/kura.rs [hunk 13]
```diff
@@ -180,126 +189,249 @@ impl Default for Mode {
...
+                if block_height <= *max_base {
+                    return Err(eyre::eyre!("An attempt to random-access persistent storage (file except for the last one queued)."));
+                }
+                if (block_height.get() - max_base.get()) < blocks_per_file.get() {
+                    return Ok(max_base.to_string());
+                }
+                Ok(block_height.to_string())
+            }
+            None => Ok(String::from(INITIAL_DATA_FILE)),
+        }
```
Comment (i1i1) [inline iroha/src/kura.rs:231]:
```suggestion
        let max_base = match storage_indices.last() {
            Some(max_base) => *max_base,
            None => return Ok(INITIAL_DATA_FILE.to_owned()),
        };
        if block_height <= *max_base {
            return Err(eyre::eyre!("An attempt to random-access persistent storage (file except for the last one queued)."));
        }
        if (block_height.get() - max_base.get()) < blocks_per_file.get() {
            return Ok(max_base.to_string());
        }
        Ok(block_height.to_string())
```

File: iroha/src/kura.rs [hunk 14]
```diff
@@ -180,126 +189,249 @@ impl Default for Mode {
...
     /// Write block as file onto the disk.
     #[iroha_futures::telemetry_future]
     pub async fn write(&self, block: &VersionedCommittedBlock) -> Result<Hash> {
-        //filename is its height
-        let path = self.get_block_path(block.header().height);
-        let mut file = File::create(path)
+        let path = self
+            .get_block_path(
+                NonZeroU64::new(block.header().height)
+                    .ok_or_else(|| eyre::eyre!("Block height is zero."))?,
```
Comment (i1i1) [inline iroha/src/kura.rs:253]: @eadventurous may be we can change height in header to `NonZeroU64`?

File: iroha/src/kura.rs [hunk 15]
```diff
@@ -180,126 +189,249 @@ impl Default for Mode {
...
+            .open(path)
             .await
             .wrap_err("Failed to open storage file.")?;
         let hash = block.hash();
         let serialized_block: Vec<u8> = block.encode_versioned()?;
+        #[allow(clippy::cast_possible_truncation)]
+        let block_size = serialized_block.len() as u32;
+        file.write_all(&block_size.to_ne_bytes())
+            .await
+            .wrap_err("Failed to write to storage file.")?;
```
Comment (i1i1) [inline iroha/src/kura.rs:268]: I guess [`AsyncWriteExt::write_u32_le`](https://docs.rs/tokio/1.12.0/tokio/io/trait.AsyncWriteExt.html#method.write_u32_le) suits better here.

**Review by i1i1 — CHANGES_REQUESTED** [review]

File: iroha/src/kura.rs [hunk 16]
```diff
@@ -180,126 +189,249 @@ impl Default for Mode {
...
+    Ok(Some((msg, tail)))
+}
+
+/// Returns sorted Vec of datafiles base heights as u64
+///
+/// # Errors
+/// Will fail on filesystem access error
+///
+async fn sorted_storage_files_base_indices(path: &Path) -> Result<Vec<NonZeroU64>> {
+    let mut bases = Vec::new();
```
Comment (i1i1) [inline iroha/src/kura.rs:354]: Minor issue: You can use `BTreeSet`, it will automatically sort everything for you at insertion.

Comment (i1i1) [reply iroha/src/kura.rs:354]: Also you can remove `sorted` from name afterwards

File: iroha/src/kura.rs [hunk 17]
```diff
@@ -180,126 +189,249 @@ impl Default for Mode {
...
-        let mut height = 1;
-        let mut blocks = Vec::new();
-        while let Ok(block) = self.read(height).await {
-            blocks.push(block);
-            height += 1;
+/// Returns deserialized blocks from specific file
+async fn read_all_from_storage_file(path: &Path) -> Result<Vec<VersionedCommittedBlock>> {
+    let mut blocks: Vec<VersionedCommittedBlock> = Vec::new();
+    let mut file = File::open(&path).await.wrap_err("No file found.")?;
+    let metadata = metadata(&path).await.wrap_err("Unable to read metadata.")?;
```
Comment (i1i1) [inline iroha/src/kura.rs:314]: Could you import `metadata` also as `fs::metadata`?

File: iroha/src/kura.rs (hunk 14 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:252]: It would save me a whole lot of trouble. Is it a breaking change? @eadventurous

Comment (eadventurous) [reply iroha/src/kura.rs:252]: This should be ok, just pay attention to properly handle it in Sumeragi during block production

File: iroha/src/kura.rs (hunk 16 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:353]: > Also you can remove `sorted` from name afterwards

You think? Should interface not document itself? Or will there be a corresponding trait on return type?

File: iroha/src/kura.rs (hunk 17 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:313]: > Could you import `metadata` also as `fs::metadata`?

std::fs or tokio::fs?

File: iroha/src/kura.rs (hunk 14 above)
Comment (s8sato) [reply iroha/src/kura.rs:252]: It looks good to me
as 0-height has special meaning
and can be expressed as `None::<NonZero>`

File: iroha/src/kura.rs (hunk 17 above)
Comment (i1i1) [reply iroha/src/kura.rs:313]: `tokio::fs`

File: iroha/src/kura.rs (hunk 15 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:267]: I don't see a way to apply write_u32_le to File directly - only to intermediate buffer. What's the point then?

Comment (i1i1) [reply iroha/src/kura.rs:276]: I mean you write length to file, you can do it by `write_u32_le`.

File: iroha/src/kura.rs [hunk 18]
```diff
@@ -180,126 +196,250 @@ impl Default for Mode {
...
+        let path = self
+            .get_block_path(
+                NonZeroU64::new(block.header().height)
+                    .wrap_err("An attempt to write a zero-height block.")?,
+            )
+            .await?;
+        let mut file = fs::OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(path)
```
Comment (i1i1) [inline iroha/src/kura.rs:267]: Could you wrap it with `bufwriter`? As for every write request it will make a future

File: iroha/src/kura.rs [hunk 19]
```diff
@@ -180,126 +196,250 @@ impl Default for Mode {
...
             .await
             .wrap_err("Failed to open storage file.")?;
         let hash = block.hash();
         let serialized_block: Vec<u8> = block.encode_versioned()?;
+        #[allow(clippy::cast_possible_truncation)]
+        let block_size = serialized_block.len() as u32;
+        file.write_all(&block_size.to_ne_bytes())
+            .await
+            .wrap_err("Failed to write to storage file.")?;
```
Comment (i1i1) [inline iroha/src/kura.rs:276]:
```suggestion
        #[allow(clippy::cast_possible_truncation)]
        let block_size = serialized_block.len() as u32;
        file.write_u32_le(block_size).await.wrap_err("Failed to write to storage file.")?;
```

File: iroha/src/kura.rs [hunk 20]
```diff
@@ -180,126 +196,250 @@ impl Default for Mode {
...
         }
-        iroha_logger::info!("Read {} blocks from block store.", blocks.len());
-        blocks
+        blocks.push(block);
+        buffer = tail.to_vec();
+    }
+    Ok(blocks)
+}
+
+fn pop_from_buffer(mut buffer: &mut [u8]) -> Result<Option<(&[u8], &[u8])>> {
```
Comment (i1i1) [inline iroha/src/kura.rs:345]: Could you read using streams, without buffering whole file?

File: iroha/src/kura.rs [hunk 21]
```diff
@@ -180,126 +196,250 @@ impl Default for Mode {
...
+    if buffer.is_empty() {
+        return Ok(None);
     }
+
+    let (int_bytes, rest) = buffer.split_at_mut(4);
+    buffer = rest;
+    let block_size = u32::from_ne_bytes((*int_bytes).try_into()?);
```
Comment (i1i1) [inline iroha/src/kura.rs:352]:
```suggestion
    let block_size = u32::from_le_bytes((*int_bytes).try_into()?);
```

File: iroha/src/kura.rs [hunk 22]
```diff
@@ -180,126 +196,248 @@ impl Default for Mode {
...
+    ///
+    pub async fn get_block_path(&self, block_height: NonZeroU64) -> Result<PathBuf> {
+        let filename =
+            BlockStore::get_block_filename(&self.path, block_height, self.blocks_per_file)
+                .await
+                .wrap_err("Failed to fetch storage name for block")?;
+        Ok(self.path.join(filename))
     }
 
     /// Write block as file onto the disk.
```
Comment (eadventurous) [inline iroha/src/kura.rs:255]: This comment needs to be updated I think

File: iroha/src/kura.rs [hunk 23]
```diff
@@ -180,126 +196,248 @@ impl Default for Mode {
...
+            let mut read_blocks = read_all_from_storage_file(&path)
+                .await
+                .wrap_err("Failed to deserialize blocks.")?;
+            let prev_block = read_blocks
+                .first()
+                .wrap_err(format!(
+                    "No blocks were found in file {}. Data storage might be corrupted.",
+                    path.to_string_lossy()
+                ))?
+                .header()
```
Comment (eadventurous) [inline iroha/src/kura.rs:297]: In all of the cases where you return errors - you return a boxed string with message. It is ok for high level functions that just forward error up the call stack. But BlockStore is a more low level (library structure) I would say, and we might want to handle some of this errors in the future.

So I would suggest having `BlockStoreError` enum with all variants that you have here. You might use `thiserror` for this. And then for this function for example you would return `Result<Vec<VersionedCommittedBlock>, BlockStoreError>`.

Though this is not critical now I guess if you don't handle this errors anywhere on top. But might be good for the future.

**Comment by eadventurous** [comment]
As this PR introduces some new edge cases - I would suggest to also increase the number of tests and test some of the new functions and particular scenarios.

Negative tests might be especially good. And for this the suggestion to use enum for errors would be useful.

File: iroha/src/kura.rs (hunk 23 above)
Comment (i1i1) [reply iroha/src/kura.rs:297]: Also maybe a good idea will be at least to do `wrap_err_with(|| ...)`

File: iroha/src/kura.rs [hunk 24]
```diff
@@ -180,126 +196,248 @@ impl Default for Mode {
...
+            Some(max_base) => *max_base,
+            None => return Ok(INITIAL_DATA_FILE.to_owned()),
+        };
+        if block_height <= max_base {
+            return Err(eyre::eyre!("An attempt to random-access persistent storage (file except for the last one queued)."));
+        }
+        if (block_height.get() - max_base.get()) < blocks_per_file.get() {
+            return Ok(max_base.to_string());
+        }
+        Ok(block_height.to_string())
```
Comment (s8sato) [inline iroha/src/kura.rs:239]: Just a refactoring. It's up to you to adopt
```suggestion
        if let Some(max_base) = storage_indices.iter().last() {
            if *max_base < block_height {
                if block_height.get() - max_base.get() < blocks_per_file.get() {
                    return Ok(max_base.to_string());
                }
                return Ok(block_height.to_string());
            }
            return Err(eyre::eyre!("An attempt to random-access persistent storage (file except for the last one queued)."));
        }
        Ok(INITIAL_DATA_FILE.to_owned())
```

**Review by s8sato — COMMENTED** [review]
Add 1 comment

File: iroha/src/kura.rs (hunk 24 above)
Comment (eadventurous) [reply iroha/src/kura.rs:239]: Both styles are ok, as long as the code is readable(nesting is not too deep, etc.). We don't have a preference in our style guide, so this should probably stay up to the PR author.

Comment (i1i1) [reply iroha/src/kura.rs:239]: Here probably both styles are okay, but on bigger functions it is hard to go through nested code. I agree that we dont have preferences, but style is also a discussable thing.

File: iroha/src/kura.rs [hunk 25]
```diff
@@ -180,126 +196,248 @@ impl Default for Mode {
...
+    let (msg, tail) = buffer.split_at_mut(block_size as usize);
+    Ok(Some((msg, tail)))
+}
+
+/// Returns sorted Vec of datafiles base heights as u64
+///
+/// # Errors
+/// Will fail on filesystem access error
+///
+async fn storage_files_base_indices(path: &Path) -> Result<BTreeSet<NonZeroU64>> {
```
Comment (GaroRobe) [inline iroha/src/kura.rs:361]: For some reason it broke :(
I don't see a reason why - my best guess is something in uses, but I could be wrong.

File: iroha/src/kura.rs [hunk 26]
```diff
@@ -180,126 +196,248 @@ impl Default for Mode {
...
+                .first()
+                .wrap_err(format!(
+                    "No blocks were found in file {}. Data storage might be corrupted.",
+                    path.to_string_lossy()
+                ))?
+                .header()
+                .height;
+            if let Some(last_block) = blocks.last() {
+                let new_block = last_block.header().height;
+                if prev_block != new_block + 1 {
```
Comment (GaroRobe) [inline iroha/src/kura.rs:301]: This whole block is hairy at the moment.
I tried my best to explode everything I do type-by-type, and even went from map().collect() to for_each(forward()), but still cannot untangle how it should be done properly.

Comment (GaroRobe) [reply iroha/src/kura.rs:301]: Also, for the time being I ignore error handling - it's a thing to account for as well.

File: iroha/src/lib.rs [hunk 27]
```diff
@@ -311,7 +312,7 @@ const GENESIS_PATH: &str = "genesis.json";
 pub struct Arguments {
     /// Set this flag on the peer that should submit genesis on the network initial start.
     #[cfg_attr(feature = "cli", structopt(long))]
-    pub submit_genesis: bool,
+    pub submit_genesis: u8,
```
Comment (i1i1) [inline iroha/src/lib.rs:315]: Why did it became `u8`?

File: iroha/src/kura.rs [hunk 28]
```diff
@@ -180,126 +193,242 @@ impl Default for Mode {
...
                 .to_owned();
             Ok(())
         }
+
+        /// Set `blocks_per_storage_file` parameter - will owerwrite the existing one
+        ///
+        pub fn blocks_per_storage_file(&mut self, size: NonZeroU64) {
+            self.blocks_per_storage_file = size;
+        }
+    }
```
Comment (i1i1) [inline iroha/src/kura.rs:421]: Why not to make field public?

File: iroha/src/kura.rs (hunk 24 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:236]: I'll pick one with less nesting - it's easier to read IMO.

File: iroha/src/lib.rs (hunk 27 above)
Comment (GaroRobe) [reply iroha/src/lib.rs:315]: It should not, but CLI was complaining about bool parameter, that's why I made it u8. Could you give me a better option?

File: iroha/src/kura.rs (hunk 28 above)
Comment (GaroRobe) [reply iroha/src/kura.rs:421]: I did. Actually, it is.
